### ISSUE TITLE
clang-tidy wrongly reports bugprone-macro-parentheses on non-arguments

### ISSUE THREAD

**Issue by j-jorge** [llvm/llvm-project#89921]
Given the following code :
```c++
#include <cstdio>

#ifndef FOO
#define FOO a-b
#endif

int main()
{
#define STR0(x) #x
#define STR(x) STR0(x)
  printf("%s\n", STR(FOO));
}
```
A run of `clang-tidy --checks=bugprone-macro-parentheses` will complain about the need of replacing `a` and `b` by `(a)` and `(b)` ([Compiler Explorer demo](https://godbolt.org/z/ETdvP75av)) :
```
warning: macro replacement list should be enclosed in parentheses [bugprone-macro-parentheses]
[<source>:4:14: warning: macro replacement list should be enclosed in parentheses [bugprone-macro-parentheses]](javascript:;)
    4 | #define FOO a-b
```

I believe this is a false positive. Since none of `a` or `b` are arguments of the macro they should not be considered by the check.

My actual use case is similar to this code. I have a preprocessor definition coming from the command line (`-DFOO=some-string-with-dashes`) which I turn into a C-string in the code via the `STR()` macro. The workaround is to drop the `STR` macro and put the quotes in the command-line, but I would have preferred no workaround :)

**Comment by LegalizeAdulthood** [comment]
This came from a discussion on slack, so I will repeat my comment from there:

Ah!  Now that I see it as a `#define`, it makes sense.  This isn't a quoted string expansion, it's an expansion into multiple tokens (namely `a`, `-`, and `b`).  Furthermore, those tokens that are valid identifiers could be further macro expansions, so this is not a false positive, this is a genuine case where you need the parenthesis.  Consider the following expression:
```
double a = 1.0/FOO;
```
Without parenthesis the expansion is:
```
double a = 1.0/a-b; // == (1.0/a) - b
```
This is probably not what you intended and therefore the parenthesis are properly suggested.

**Comment by j-jorge** [comment]
Closing because there is an actual use case, as described by @LegalizeAdulthood, and because there is a workaround.
